LSM review & documentation

sentientwaffle opened the pull request:
Specific changes:

- `Groove`: Use `mem.indexOf` instead of `mem.containsAtLeast` — the latter returns `true` even if the hash was reused.
- `Groove.deinit`/`PostedGroove.deinnit`: Remove some assertions — `deinit` may be called while a "join" operation is in progress.
- `SegmentedArray`: Make `Cursor` "generic", so that cursors from different `SegmentedArray`s are not interchangeable. As a bonus, this makes the `SegmentedArray` that a cursor refers to somewhat self-documenting.
- `Tree`: Panic if `checkpoint` is called on an invalid beat, rather than just silently not checkpointing.
- `Grid`: Add `BlockOperation` for additional block type verification.

Misc
- Move `commit_count_max_*` into `config.zig`
- Move `storage.tick()` from `Forest` to `main`. `Storage` doesn't belong to the `Forest`.
- Add `prefetch_count_max` options to `Groove`/`StateMachine`.

Not included above:
- Many new (or tighter) assertions.
- Additional and updated documentation
- Formatting
- Remove dead code.
- Move `storage.tick()` out of `Forest` into `main`. (`Forest` doesn't own the storage). Also, simulator was already calling `storage.tick()`.

sentientwaffle committed:
LSM: Code review (docs & fixes)

Specific changes:

Groove
- Use `mem.indexOf` instead of `mem.containsAtLeast` — the latter returns `true` even if the hash was reused.
- `deinit`: Remove some assertions — `deinit` may be called while a "join" operation is in progress.
SegmentedArray
- Make `Cursor` "generic", so that cursors from different `SegmentedArray`s are not interchangeable. As a bonus, this makes the `SegmentedArray` that a cursor refers to somewhat self-documenting.
- `Tree`: Panic if `checkpoint` is called on an invalid beat, rather than just silently not checkpointing.
Grid
- Add `BlockOperation` to label & verify the block type.
Misc
- Move `commit_count_max_*` into `config.zig`
- Move `storage.tick()` from `Forest` to `main`. `Storage` doesn't belong to the `Forest`.

Not included above:
- Many new (or tighter) assertions.
- Additional and updated documentation
- Formatting
- Remove dead code.
- Move `storage.tick()` out of `Forest` into `main`. (`Forest` doesn't own the storage). Also, simulator was already calling `storage.tick()`.

sentientwaffle committed:
LSM: Add prefetch_count_max options

jorangreef commented on src/config.zig:
```diff
@@ -65,6 +66,25 @@ pub const transfers_max = switch (deployment_environment) {
+// TODO Move these to a separate "internal computed constants" file.
+// 128 == @sizeOf(Header)
+const message_body_accounts = @divFloor(message_size_max - 128, @sizeOf(tigerbeetle.Account));
+const message_body_transfers = @divFloor(message_size_max - 128, @sizeOf(tigerbeetle.Transfer));
```
```suggestion
const message_body_count_max_accounts = @divFloor(message_body_size_max, @sizeOf(tigerbeetle.Account));
const message_body_count_max_transfers = @divFloor(message_body_size_max, @sizeOf(tigerbeetle.Transfer));
```

It would be good to introduce `message_body_size_max`, so that we don't have to otherwise subtract `@sizeOf(vsr.Header)` twice, and so that we don't have to use `128` as a magic number, which would otherwise get out of sync if the header size is ever changed.

jorangreef commented on src/config.zig:
```diff
@@ -65,6 +66,25 @@ pub const transfers_max = switch (deployment_environment) {
+pub const commit_count_max_posted = message_body_transfers;
+
+// *2 to fetch a transfer's debit/credit accounts.
+pub const prefetch_count_max_accounts = 2 * message_body_accounts;
```
```suggestion
// StateMachine.create_transfer()/post_or_void_pending_transfer() both lookup 2 accounts for every transfer.
pub const prefetch_count_max_accounts = 2 * message_body_transfers;
```
We also need to take the max of `2 * message_body_transfers` and `message_body_accounts` just to be safe, in case the number of accounts per message ever changes.

jorangreef commented on src/config.zig:
```diff
@@ -65,6 +66,25 @@ pub const transfers_max = switch (deployment_environment) {
+/// See `Groove.init`'s `commit_count_max` for more detail.
+// *2 because creating a transfer will update 2 accounts.
+pub const commit_count_max_accounts = 2 * message_body_accounts;
+// *2 because creating a post/void-transfer queries the post/void transfer and the pending transfer.
```
> creating a post/void-transfer queries the post/void transfer

`commit_count` is not about what's queried so much as how many objects/values will be committed/inserted.

jorangreef replied:
i.e. This comment seems to belong more to `prefect_count_max_transfers` below?

jorangreef replied:
How does this value interact also with indexes for accounts? For example, if a transfer mutates the balances of 2 accounts, then I think this could actually turn into 4 commits (inserts/updates) especially for account secondary indexes? i.e. 1 to remove the secondary index for account A's old value, 1 to insert the secondary index for account A's new value, and likewise for account B?

sentientwaffle replied:
`commit_count_max_transfers` doesn't impact the accounts (or account indexes).

The `*2` explanation was a little speculative on my part — in `main` it is already [there](https://github.com/tigerbeetledb/tigerbeetle/blob/68ad415cdc5d1c1c696463577bb1fcda0f021e82/src/state_machine.zig#L100-L103), (but there wasn't one on `accounts`).

But it sounds like it should be:

```
pub const commit_count_max_accounts = 4 * message_body_count_max_accounts;
pub const commit_count_max_transfers = message_body_count_max_transfers;
```
is that right?

jorangreef replied:
> commit_count_max_transfers doesn't impact the accounts (or account indexes).

I think it does though, since each transfer committed may change the balances for 2 accounts (with up to 2 commits for each of those account balances, 1 to remove the old balance index, 1 to insert the new balance index)?

jorangreef commented on src/config.zig:
```diff
@@ -235,8 +255,11 @@ pub const block_count_max = @divExact(16 * 1024 * 1024 * 1024 * 1024, block_size
 pub const lsm_trees = 30;
 
 /// The number of levels in an LSM tree.
+/// Higher levels increases read amplification, but increase total storage capacity.
```
```suggestion
/// A higher number of levels increases read amplification, as well as total storage capacity.
```

jorangreef reviewed (commented):
Getting back into it!

jorangreef commented on src/config.zig:
```diff
@@ -235,8 +255,11 @@ pub const block_count_max = @divExact(16 * 1024 * 1024 * 1024 * 1024, block_size
+/// Higher levels increases read amplification, but increase total storage capacity.
 pub const lsm_levels = 7;
 
+/// The number of tables at level i (0≤i<lsm_levels) is `pow(lsm_growth_factor, i+1)`.
```
```suggestion
/// The number of tables at level i (0 ≤ i < lsm_levels) is `pow(lsm_growth_factor, i+1)`.
/// See also `tree.table_count_max_for_level()`.
```

jorangreef commented on src/config.zig:
```diff
@@ -235,8 +255,11 @@ pub const block_count_max = @divExact(16 * 1024 * 1024 * 1024 * 1024, block_size
 pub const lsm_levels = 7;
 
+/// The number of tables at level i (0≤i<lsm_levels) is `pow(lsm_growth_factor, i+1)`.
+/// A higher growth factor increases write amplification, but decreases read amplification.
```
```suggestion
/// A higher growth factor increases write amplification (by increasing the number of tables in
/// level B that overlap a table in level A in a compaction), but decreases read amplification (by
/// reducing the height of the tree and thus the number of levels that must be probed). Since read
/// amplification can be optimized more easily (with filters and caching), we target a growth
/// factor of 8 for lower write amplification rather than the more typical growth factor of 10.
```

jorangreef commented on src/config.zig:
```diff
@@ -341,8 +364,20 @@ comptime {
+
+    assert(commit_count_max_accounts > 0);
+    assert(commit_count_max_transfers > 0);
+    assert(commit_count_max_posted > 0);
```
We could move this also into the state machine.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -1,3 +1,32 @@
+//! Compaction moves and merges a table's values into the next level.
```
```suggestion
//! Compaction moves or merges a table's values into the next level.
```

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -1,3 +1,32 @@
+//!
+//! 5. Update the old level-B tables in the Manifest with their new `snapshot_max`.
+//!
+//! 6. Insert the new level-B tables into the Manifest.
```
```suggestion
//! 4. Write the sort-merge iterator into a sequence of new tables on disk.
//!
//! 5. Update the old level-B tables in the Manifest with their new `snapshot_max` so that they
//!    become invisible to subsequent read transactions.
//!
//! 6. Insert the new level-B tables into the Manifest.
```

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -1,3 +1,32 @@
+//! Compaction moves and merges a table's values into the next level.
+//!
+//!
+//! Compaction overview:
```
Thanks for this terrific overview!

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -90,6 +119,7 @@ pub fn CompactionType(
         data: BlockWrite,
 
         remove_level_a: ?*const TableInfo = null,
+        // Batch manifest updates/inserts.
```
```suggestion
        // Batch updates/inserts to the in-memory manifest, to amortize index updates there.
```

sentientwaffle replied:
Batching will these updates/inserts will be removed in the upcoming manifest level refactor.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -374,7 +409,7 @@ pub fn CompactionType(
             assert(compaction.io_pending == 0);
 
-            // Consume the callback and invoke it one finished updating state below.
+            // Consume the callback and invoke it after updating state below.
```
```suggestion
            // Consume the callback and invoke it after updating the state below.
```

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -473,6 +509,7 @@ pub fn CompactionType(
                 if (compaction.table_builder.filter_block_empty()) {
                     assert(compaction.drop_tombstones);
                     assert(tombstones_dropped > 0);
+                    assert(!compaction.data.ready);
```
Nice!

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -515,6 +555,7 @@ pub fn CompactionType(
         }
 
         fn io_callback(compaction: *Compaction) void {
+            assert(compaction.status == .compacting);
```
@kprotty would you please double-check that we can only reach `io_callback()` if the compaction status is `.compacting`?

sentientwaffle replied:
This branch needs to be rebased onto @kprotty's compaction refactor, so some of these changes are probably broken or no-longer applicable anyway.

jorangreef commented on src/lsm/composite_key.zig:
```diff
@@ -27,6 +27,8 @@ pub fn CompositeKey(comptime Field: type) type {
             comptime {
                 assert(@sizeOf(Value) == @sizeOf(Field) * 2);
                 assert(@alignOf(Value) == @alignOf(Field));
+                // Assert that there is no implicit padding in the struct.
```
```suggestion
```

Since we'll be doing this more and more frequently in so many places.

jorangreef commented on src/lsm/composite_key.zig:
```diff
@@ -38,6 +40,8 @@ pub fn CompositeKey(comptime Field: type) type {
         comptime {
             assert(@sizeOf(Self) == @sizeOf(Field) * 2);
             assert(@alignOf(Self) == @alignOf(Field));
+            // Assert that there is no implicit padding in the struct.
```
```suggestion
```

jorangreef commented on src/lsm/forest.zig:
```diff
@@ -24,6 +25,16 @@ pub fn ForestType(comptime Storage: type, comptime groove_config: anytype) type
             },
         };
+
+        groove_options_fields = groove_options_fields ++ [_]std.builtin.TypeInfo.StructField{
```
@kprotty would you please review all this comptime work w.r.t. `groove_options_fields`?

jorangreef commented on src/lsm/forest.zig:
```diff
@@ -35,33 +46,21 @@ pub fn ForestType(comptime Storage: type, comptime groove_config: anytype) type
         .Struct = .{
             .layout = .Auto,
-            .fields = all_groove_options_fields,
+            .fields = groove_options_fields,
```
A good improvement on the name.

jorangreef commented on src/lsm/forest.zig:
```diff
@@ -35,33 +46,21 @@ pub fn ForestType(comptime Storage: type, comptime groove_config: anytype) type
+    //    GrooveOptions,
+    //    null,
+    //);
+
```
Can this be removed?

sentientwaffle replied:
Yes, I'll remove it. I was experimenting with a cleaner implementation for this metaprogramming, but eventually decided the `Groove` type should by able to have different `Options` types, so `std.enums.EnumFieldStruct` isn't applicable.

jorangreef commented on src/lsm/forest.zig:
```diff
@@ -86,7 +85,7 @@ pub fn ForestType(comptime Storage: type, comptime groove_config: anytype) type
             grid: *Grid,
             node_count: u32,
-            // (e.g.) .{ .transfers = .{ cache_size = 128, com_count_max = n }, .accounts = same }
+            // (e.g.) .{ .transfers = .{ .cache_size = 128, .commit_count_max = n }, .accounts = same }
```
✔️

jorangreef commented on src/lsm/forest.zig:
```diff
@@ -107,15 +106,15 @@ pub fn ForestType(comptime Storage: type, comptime groove_config: anytype) type
-                const groove_options: GrooveOptions = @field(all_groove_options, groove_field.name);
                 const groove = &@field(grooves, groove_field.name);
+                const Groove = @TypeOf(groove.*);
+                const groove_options: Groove.Options = @field(all_groove_options, groove_field.name);
```
While we're making changes here, would be great also to have `grooves_options` rather than `all_groove_options`.

jorangreef replied:
...since “all groove options” could be misread as ”all the options for a groove”.

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -12,6 +12,32 @@ const SetAssociativeCache = @import("set_associative_cache.zig").SetAssociativeC
 const log = std.log.scoped(.grid);
 
+/// A block's type is implicitly determined by how its address is stored (e.g. in the index block).
+/// BlockOperation is an additional check that a block has the expected type on read.
```
Let's call it `BlockType` too, since that's what it is (and our comment on line 15 also describes it like that: “A block's type”), even though we store it in `header.operation` (and later, we'll also be updating `vsr.Header` to have different field names according to the `command`).

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -12,6 +12,32 @@ const SetAssociativeCache = @import("set_associative_cache.zig").SetAssociativeC
+    filter = 3,
+    data = 4,
+
+    pub inline fn from(operation_vsr: vsr.Operation) BlockOperation {
```
```suggestion
    pub inline fn from(vsr_operation: vsr.Operation) BlockType {
```

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -12,6 +12,32 @@ const SetAssociativeCache = @import("set_associative_cache.zig").SetAssociativeC
+        return @intToEnum(BlockOperation, @enumToInt(operation_vsr));
+    }
+
+    pub inline fn operation(operation_block: BlockOperation) vsr.Operation {
```
```suggestion
    pub inline fn operation(block_type: BlockType) vsr.Operation {
```

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -12,6 +12,32 @@ const SetAssociativeCache = @import("set_associative_cache.zig").SetAssociativeC
+    }
+};
+
+/// The Grid provides access to on-disk blocks (blobs of `block_size` bytes).
+/// Each block is identified by an "address" (`u64`, beginning at 1).
```
So much clearer to help people get into Grid, also the “beginning at 1”.

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -412,11 +442,14 @@ pub fn GridType(comptime Storage: type) type {
                 while (iop.reads.pop()) |read| {
                     assert(read.address == address);
                     assert(read.checksum == checksum);
+                    assert(read.operation == BlockOperation.from(header.operation));
```
Would be great to also improve the logging for the else branch below, i.e. to compare the block type that was expected vs what was found.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -160,13 +160,13 @@ pub fn GrooveType(
 
     // Generate index LSM trees from the struct fields.
-    inline for (std.meta.fields(Object)) |field| {
+    for (std.meta.fields(Object)) |field| {
```
What is the reason for dropping the `inline`? Do we not need it to avoid comptime bugs? CC @kprotty

sentientwaffle replied:
`std.meta.fields(Object)`'s return value is already comptime, so I think inlining the loop isn't necessary. If there are bugs that `inline` would avoid though, please lmk.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -273,17 +273,17 @@ pub fn GrooveType(
         const IndexTree = @TypeOf(@field(@as(IndexTrees, undefined), field.name));
         const hash: []const u128 = &.{IndexTree.hash};
 
-        assert(std.mem.containsAtLeast(u128, hashes, 0, hash));
```
Good catch! This was effectively a no-op, with no search being done, and nothing being checked in the assertion, since `containsAtLeast` was doing an early return seeing that we didn't expect any matches.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -397,41 +397,47 @@ pub fn GrooveType(
+            /// TODO Improve unit in this name to make more clear what should be passed.
+            /// For example, is this a size in bytes or a count in objects? It's a count in objects,
+            /// but the name poorly reflects this.
+            cache_size: u32,
```
```suggestion
            cache_cardinality_max: u32,
```

sentientwaffle replied:
What about "members" or "entries" or "objects" in place of "cardinality"?

jorangreef replied:
`entries` is good. `cardinality` is almost the perfect term though. It's the count of elements in a set.

sentientwaffle replied:
Good point, I'll update it. Just btw, my thinking with `entries` and `objects` was what we could also change `commit_count_max` to `commit_objects_max`, to clarify that it isn't a limit on the number of commits. (Likewise for `prefetch_count_max`).

jorangreef replied:
Ah, good point. Shall we do `cache_entries_max`, `commit_entries_max` etc. then?

Rationale being that perhaps `entries` is better than `objects` since secondary indexes don't store objects.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -397,41 +397,47 @@ pub fn GrooveType(
 
+        pub const Options = struct {
+            /// The cache size is meant to be computed based on the left over available memory
+            /// that tigerbeetle was given to allocate from CLI arguments.
```
I think we'll change this strategy to let the user specify at the CLI how much RAM they want to give to the caches, then we work out the `cardinality_max` from there, knowing the size of the types.

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -397,41 +397,47 @@ pub fn GrooveType(
+            /// but the name poorly reflects this.
+            cache_size: u32,
+            /// In general, the commit count max for a field, depends on the field's object,
+            /// how many objects might be changed by a batch:
```
```suggestion
            /// how many objects might be inserted/updated/removed by a batch:
```

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -397,41 +397,47 @@ pub fn GrooveType(
-            commit_count_max: u32,
+            options: Options,
         ) !Groove {
-            // Cache is dynamically allocated to pass a pointer into the Object tree.
```
Good catch!

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -515,10 +518,6 @@ pub fn GrooveType(
         }
 
         pub fn deinit(groove: *Groove, allocator: mem.Allocator) void {
-            assert(groove.join_op == null);
-            assert(groove.join_pending == 0);
-            assert(groove.join_callback == null);
```
@kprotty Is it safe to `deinit()` while these joins are not complete?

sentientwaffle replied:
The VOPR was hitting this (frequently); it calls `deinit()` to clean up after a simulated replica crash.

jorangreef replied:
Is it safe to do that though?

jorangreef commented on src/lsm/k_way_merge.zig:
```diff
@@ -25,7 +25,12 @@ pub fn KWayMergeIterator(
 
         context: *Context,
 
-        /// Sorted array of keys, with each key representing the next key in each stream.
```
Another good catch! The array itself is not sorted, it only represents the earliest key in each stream.

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -36,8 +38,8 @@ pub fn TableInfoType(comptime Table: type) type {
-        key_min: Key,
-        key_max: Key,
+        key_min: Key, // inclusive
+        key_max: Key, // inclusive
```
```suggestion
        key_min: Key, // Inclusive.
        key_max: Key, // Inclusive.
```

jorangreef commented on src/lsm/manifest.zig:
```diff
@@ -239,7 +241,7 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
             assert(manifest.open_callback != null);
 
-            // Insert all left-over pushed open tables into the ManifestLevels.
+            // Flush left-over buffered open tables into the ManifestLevels.
```
```suggestion
            // Flush leftover buffered open tables into the ManifestLevels.
```

jorangreef commented on src/lsm/manifest_level.zig:
```diff
@@ -1,3 +1,51 @@
+//! * The sides of each rectangle depict:
+//!   * left:   table.key_min (the diagram is inclusive, and the table.key_min is inclusive)
+//!   * right:  table.key_max (the diagram is EXCLUSIVE, but the table.key_max is INCLUSIVE)
+//!   * bottom: table.snapshot_min (exclusive)
+//!   * top:    table.snapshot_max (exclusive)
```
@sentientwaffle and @kprotty the three of us should get together to talk through and draw out a few examples of how snapshots are assigned during compaction, which previous measures the underlying values in the tables come from, and how these values relate to older read transactions that might have been able to see them.

We want to ensure that read transactions always see what they should see, whether these values are later compacted, or compacted in the same measure as the read transaction, whether the read comes out of the immutable table or later comes from disk etc. because read snapshots will eventually be long-lived and may query the same values even as they are shifting between immutable table and on disk tables.

For example, I'm not sure, but I have a feeling that `snapshot_min/max` might actually be inclusive.

sentientwaffle replied:
:+1: That would be very helpful; especially in understanding how long-lived snapshots will work.

jorangreef commented on src/lsm/manifest_level.zig:
```diff
@@ -395,8 +452,8 @@ pub fn ManifestLevelType(
-            key_min: Key,
-            key_max: Key,
+            key_min: Key, // inclusive
+            key_max: Key, // inclusive
```
```suggestion
            key_min: Key, // Inclusive.
            key_max: Key, // Inclusive.
```

jorangreef commented on src/lsm/manifest_level.zig:
```diff
@@ -546,12 +603,12 @@ pub fn ManifestLevelType(
             }
 
-            const key = switch (direction) {
+            const key_start = switch (direction) {
```
This might not be the start key, because the binary search might land in the middle of a cluster of keys that are all the same (with different snapshots). We'll later move left or right in `iterator_stary_boundary()` to find the earliest/latest starting key.

jorangreef commented on src/lsm/manifest_log.zig:
```diff
@@ -31,8 +31,15 @@ const vsr = @import("../vsr.zig");
 
+/// ManifestLog block schema:
+/// │ vsr.Header                  │
+/// │ [entry_count_max]Label      │ level index, insert|remove
```
It would be great to assert also in `block_size()` that `@sizeOf(vsr.Header) + labels_size` is aligned to TableInfo's alignment. We have this check elsewhere but it would be great to see if there are a few more places we can check it.

jorangreef commented on src/lsm/manifest_log.zig:
```diff
@@ -649,14 +676,14 @@ pub fn ManifestLogType(comptime Storage: type, comptime TableInfo: type) type {
                 TableInfo,
-                block[@sizeOf(vsr.Header) + entry_count_max ..][0..tables_size_max],
+                block[@sizeOf(vsr.Header) + labels_size_max ..][0..tables_size_max],
             )[0..entry_count_max];
         }
 
         fn tables_const(block: BlockPtrConst) *const [entry_count_max]TableInfo {
             return mem.bytesAsSlice(
                 TableInfo,
-                block[@sizeOf(vsr.Header) + entry_count_max ..][0..tables_size_max],
+                block[@sizeOf(vsr.Header) + labels_size_max ..][0..tables_size_max],
```
Much better/safer!

jorangreef commented on src/lsm/posted_groove.zig:
```diff
@@ -90,42 +90,51 @@ pub fn PostedGrooveType(comptime Storage: type) type {
+            /// However, create_transfers will put 2 accounts (8191 * 2) for every transfer, and
+            /// some of these accounts may exist, requiring a remove/put to update the index.
+            commit_count_max: u32,
+            /// The maximum number of objects that might be prefetched by a batch.
```
Let's drop these duplicate comments and point instead to the same comments in `groove.zig`.

jorangreef commented on src/lsm/segmented_array.zig:
```diff
@@ -9,11 +9,14 @@ const div_ceil = @import("../util.zig").div_ceil;
-    relative_index: u32,
-};
-
+/// A "segmented array" is an array with efficient (amortized) random-insert/remove operations.
```
```suggestion
/// A "segmented array" is an array with efficient (amortized) random-insert/remove operations.
/// Also known as an "unrolled linked list": https://en.wikipedia.org/wiki/Unrolled_linked_list
```

jorangreef commented on src/lsm/segmented_array.zig:
```diff
@@ -9,11 +9,14 @@ const div_ceil = @import("../util.zig").div_ceil;
+/// When a node empties, it is joined with a nearby node.
+///
+/// An absolute index is offset from the start of the segmented array.
+/// A relative index is offset from the start of the node.
```
```suggestion
/// A relative index is offset from the start of a node.
```

jorangreef commented on src/lsm/segmented_array.zig:
```diff
@@ -9,11 +9,14 @@ const div_ceil = @import("../util.zig").div_ceil;
-
+/// A "segmented array" is an array with efficient (amortized) random-insert/remove operations.
+///
+/// The structure consists of an arraylist of "nodes". Each node is a non-empty array of T.
```
```suggestion
/// The structure consists of an array list of "nodes". Each node is a non-empty array of T.
```

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -19,6 +19,8 @@ pub const Layout = struct {
     value_alignment: ?u29 = null,
 };
 
+/// An n-way set-associative cache may store each Value in one of n locations.
```
```suggestion
/// An n-way set-associative cache may store each Value in one of n consecutive locations.
```

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -19,6 +19,8 @@ pub const Layout = struct {
 };
 
+/// An n-way set-associative cache may store each Value in one of n locations.
+/// The Value's possible locations are determined by its Key.
```
Another way to phrase this is:

> Each Key is associated with a set of n consecutive ways (or slots) that may contain the Value.

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -86,9 +88,29 @@ pub fn SetAssociativeCache(
+
+        /// A short, partial hash of a Key, corresponding to a Value.
+        /// Because the tag is short, collisions are possible:
+        /// `tag(v₁) = tag(v₂)` does not imply `v₁ = v₂`.
```
```suggestion
        /// A short, partial hash of a Key, corresponding to a Value.
        /// Because the tag is small, collisions are possible:
        /// `tag(v₁) = tag(v₂)` does not imply `v₁ = v₂`.
        /// However, most of the time, where the tag differs, a full key comparison can be avoided.
        /// Since tags are 16-32x smaller than keys, they can also be kept hot in cache.
```

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -86,9 +88,29 @@ pub fn SetAssociativeCache(
+        ///
         counts: PackedUnsignedIntegerArray(Count),
+
+        /// Each set has a Clock: a counter that cycles between each of the set's members (i.e. ways).
```
```suggestion
        /// Each set has a Clock: a counter that cycles between each of the set's ways (i.e. slots).
```

Then it's also under the 100 column limit.

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -86,9 +88,29 @@ pub fn SetAssociativeCache(
+
+        /// Each set has a Clock: a counter that cycles between each of the set's members (i.e. ways).
+        /// The clock ensures that entries are not evicted more/less often depending on their index
+        /// within `values`.
```
```suggestion
```

We could point to the CLOCK algorithm described here: https://en.wikipedia.org/wiki/Page_replacement_algorithm

Our version is known as CLOCK Nth-Chance since we use more than 1 bit per way, to give the way more than 1 chance. The Second Chance algorithm is not actually the same (although often conflated with CLOCK) because IIRC it swaps pointers around.

There's also a nice explanation of what is essentially Nth-Chance (called RRIParoo) in FB's Kangaroo caching paper (killer paper): https://www.pdl.cmu.edu/PDL-FTP/NVM/McAllister-SOSP21.pdf

CLOCK Nth-Chance gives great guarantees for all kinds of cache thrashing scenarios and it's so simple/intuitive/tunable and efficient at the same time.

jorangreef commented on src/lsm/table.zig:
```diff
@@ -12,8 +12,47 @@ const div_ceil = @import("../util.zig").div_ceil;
 
+/// A table is a set of blocks:
+///
+/// * Index block (1)
```
```suggestion
/// * Index block (exactly 1)
```

jorangreef commented on src/lsm/table.zig:
```diff
@@ -12,8 +12,47 @@ const div_ceil = @import("../util.zig").div_ceil;
+/// A table is a set of blocks:
+///
+/// * Index block (1)
+/// * Filter blocks (at most `filter_block_count_max`)
```
```suggestion
/// * Filter blocks (at least 1, at most `filter_block_count_max`)
```

jorangreef commented on src/lsm/table.zig:
```diff
@@ -12,8 +12,47 @@ const div_ceil = @import("../util.zig").div_ceil;
+/// * Index block (1)
+/// * Filter blocks (at most `filter_block_count_max`)
+///   Each filter block summarizes the keys for several adjacent (in terms of key range) data blocks.
+/// * Data blocks (at most `data_block_count_max`)
```
```suggestion
/// * Data blocks (at least 1, at most `data_block_count_max`)
```

jorangreef commented on src/lsm/table.zig:
```diff
@@ -12,8 +12,47 @@ const div_ceil = @import("../util.zig").div_ceil;
+///
+/// * Index block (1)
+/// * Filter blocks (at most `filter_block_count_max`)
+///   Each filter block summarizes the keys for several adjacent (in terms of key range) data blocks.
```
```suggestion
///   Each filter block summarizes the keys for several adjacent (in terms of key) data blocks.
```

To get it under the 100 column limit.

jorangreef commented on src/lsm/table.zig:
```diff
@@ -12,8 +12,47 @@ const div_ceil = @import("../util.zig").div_ceil;
+/// │ vsr.Header               │
+/// │ [block_key_count + 1]Key │ Eytzinger-layout keys from a subset of the values.
+/// │ [≤value_count_max]Value  │ At least one value (no empty tables).
+/// │ […]u8{0}                 │ padding (to end of block)
```
Thanks @sentientwaffle this is so great to have.

jorangreef commented on src/lsm/table.zig:
```diff
@@ -158,15 +197,18 @@ pub fn TableType(
             var filter_blocks = 0;
-            var filter_index_size = 0;
             while (true) : (data_blocks -= 1) {
-                data_index_size = data_index_entry_size * data_blocks;
```
Surprising (!) that we were doing this before, leaking the variable outside of the scope as well.

jorangreef commented on src/lsm/table.zig:
```diff
@@ -178,14 +220,18 @@ pub fn TableType(
 
-            break :blk .{
+            break :layout .{
+                // The number of keys per data block.
```
```suggestion
                // The number of keys in the Eytzinger layout per data block.
```

sentientwaffle replied:
Oops! Great catch.

jorangreef commented on src/lsm/table_immutable.zig:
```diff
@@ -96,7 +101,7 @@ pub fn TableImmutableType(comptime Table: type) type {
                     const left_key = key_from_value(&sorted_values[i - 1]);
                     const right_key = key_from_value(&sorted_values[i]);
-                    assert(compare_keys(left_key, right_key) != .gt);
+                    assert(compare_keys(left_key, right_key) == .lt);
```
Great catch!

jorangreef commented on src/lsm/table_immutable.zig:
```diff
@@ -174,21 +179,23 @@ pub fn TableImmutableIteratorType(comptime Table: type, comptime Storage: type)
 
         pub fn tick(it: *const TableImmutableIterator) bool {
-            _ = it;
+            assert(!it.table.free);
```
This made me smile, so much better use of the `it` variable than before!

jorangreef commented on src/lsm/table_iterator.zig:
```diff
@@ -9,6 +9,7 @@ const RingBuffer = @import("../ring_buffer.zig").RingBuffer;
 const ManifestType = @import("manifest.zig").ManifestType;
 const GridType = @import("grid.zig").GridType;
 
+/// A TableIterator iterates a table's values in ascending-key order.
```
We need to improve this soon to support `direction` so that we can use `TableIterator` for ascending/descending range queries.

jorangreef commented on src/lsm/table_iterator.zig:
```diff
@@ -100,8 +101,8 @@ pub fn TableIteratorType(comptime Table: type, comptime Storage: type) type {
             grid: *Grid,
-            address: u64,
-            checksum: u128,
+            address: u64, // table index block address
```
```suggestion
            address: u64, // Table index block address.
```

jorangreef commented on src/lsm/table_iterator.zig:
```diff
@@ -100,8 +101,8 @@ pub fn TableIteratorType(comptime Table: type, comptime Storage: type) type {
-            address: u64,
-            checksum: u128,
+            address: u64, // table index block address
+            checksum: u128, // table index block checksum
```
```suggestion
            checksum: u128, // Table index block checksum.
```

jorangreef commented on src/lsm/table_mutable.zig:
```diff
@@ -59,19 +59,20 @@ pub fn TableMutableType(comptime Table: type) type {
-            const gop = table.values.getOrPutAssumeCapacity(tombstone);
-            gop.key_ptr.* = tombstone;
+            const gop = table.values.getOrPutAssumeCapacity(value.*);
+            gop.key_ptr.* = tombstone_from_key(key_from_value(value));
```
Much better!

What can we rename `gop` to? `entry`?

jorangreef replied:
Or `upsert`?

sentientwaffle replied:
Zig calls it a `GetOrPutResult`, but `result` is too general. I like `upsert` — very specific, and we don't use the term anywhere else.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -435,7 +439,8 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
         //  - (third) down beat of the measure:
         //      * assert: no compactions are currently running.
-        //      * start odd level and immutable table compactions.
+        //      * start odd level compactions if there's any tables to compact.
```
```suggestion
        //      * start odd level compactions if there are any tables to compact, and only if we must.
```

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -435,7 +439,8 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
         //      * assert: no compactions are currently running.
-        //      * start odd level and immutable table compactions.
+        //      * start odd level compactions if there's any tables to compact.
+        //      * compact immutable table if it contains any sorted values (could be empty).
```
```suggestion
        //      * compact the immutable table if it contains any sorted values (it could be empty).
```

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -865,10 +870,10 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
             assert(tree.compaction_io_pending == 0);
             assert(tree.compaction_callback == null);
 
-            // Avoid checkpointing if this is not the last beat in the compaction measure.
+            // Assert that this is the last beat in the compaction measure.
             const compaction_beat = tree.compaction_op % config.lsm_batch_multiple;
             const last_beat_in_measure = config.lsm_batch_multiple - 1;
-            if (compaction_beat != last_beat_in_measure) return callback(tree);
+            assert(last_beat_in_measure == compaction_beat);
```
@kprotty would you double-review this please?

sentientwaffle replied:
:+1: Just for extra context:

- The replica only calls `checkpoint()` at the correct tick (if it doesn't, that's a bug).
- I think this is the safer approach because otherwise the replica might call `checkpoint()`, resume from the callback, but assume (incorrectly) that the LSM actually checkpointed.

jorangreef replied:
100%

jorangreef commented on src/main.zig:
```diff
@@ -130,6 +130,12 @@ const Command = struct {
+                .commit_count_max_posted = config.commit_count_max_posted,
+                .prefetch_count_max_accounts = config.prefetch_count_max_accounts,
+                .prefetch_count_max_transfers = config.prefetch_count_max_transfers,
+                .prefetch_count_max_posted = config.prefetch_count_max_posted,
```
Great to have this flexibility!

jorangreef commented on src/main.zig:
```diff
@@ -148,6 +154,7 @@ const Command = struct {
 
         while (true) {
             replica.tick();
+            command.storage.tick();
```
Let's move this within `replica.tick()` since replica ultimately is responsible for `storage`.

sentientwaffle replied:
:thinking: Right now, `storage.tick()` just calls `io.tick()`. There is 1 storage : 1 replica ratio, but if you (for some reason) had multiple replicas in a process, they would all share an IO, and you wouldn't want to tick IO that more than once per replica tick.

(Not that you would ever actually want to run multiple replicas in a single process (outside of the simulator), I'm just trying to convey the abstractions. This is mostly in the interest of making TigerBeetle library-friendly)

Maybe replica should call `storage.tick()`, but `io.tick()` should move out of storage, to the top-level loop? It is a little weird that it is in `Storage` when the `MessageBus` also uses `IO`.

jorangreef replied:
> Maybe replica should call storage.tick(), but io.tick() should move out of storage, to the top-level loop? It is a little weird that it is in Storage when the MessageBus also uses IO.

Exactly—I think your suggestion here is the right way to solve it.

jorangreef commented on src/vsr/superblock.zig:
```diff
@@ -52,6 +52,7 @@ pub const SuperBlockSector = extern struct {
     size: u64,
 
     /// The maximum size of the data file.
+    // TODO Actually limit the file to this size.
```
Indeed. I also meant this to be here as a TODO! It's our "cgroups" for data files.

jorangreef commented on src/vsr/superblock.zig:
```diff
@@ -716,6 +716,7 @@ pub fn SuperBlockType(comptime Storage: type) type {
             if (superblock.free_set.highest_address_acquired()) |address| {
                 staging.size += address * config.block_size;
             }
+            assert(staging.size <= staging.size_max);
```
Should we also relate `staging.size` to `data_file_size_min` like we do below for `writing.size`?

jorangreef commented on src/vsr/superblock.zig:
```diff
@@ -1208,6 +1213,7 @@ pub fn SuperBlockType(comptime Storage: type) type {
             } else if (context.copy == stopping_copy_for_sequence(superblock.working.sequence)) {
                 @panic("superblock free set lost");
             } else {
+                log.debug("open: read_free_set: broken copy={}", .{ context.copy });
```
```suggestion
                log.debug("open: read_free_set: corrupt copy={}", .{ context.copy });
```

jorangreef replied:
We can't distinguish between torn writes through crash, or corruption, or updates-in-place. Figured we go with `corrupt` then to line up with our fault model.

jorangreef commented on src/vsr/superblock.zig:
```diff
@@ -1325,6 +1332,7 @@ pub fn SuperBlockType(comptime Storage: type) type {
                 if (superblock.working.manifest_size > 0) {
                     assert(superblock.manifest.count > 0);
                 }
+                // TODO Make the FreeSet encoding format not dependant on the word size.
```
Ah nice, are you thinking this would then clean up our FreeSet encoding here in SuperBlock to be the same as the Manifest?

jorangreef replied:
How do you plan to do this for FreeSet's encoding format, not to be dependent on word size?

sentientwaffle replied:
> Ah nice, are you thinking this would then clean up our FreeSet encoding here in SuperBlock to be the same as the Manifest?

Actually I'm not sure what you mean by this :thinking:.

> How do you plan to do this for FreeSet's encoding format, not to be dependent on word size?

I don't have a specific plan yet :grin: — it will probably just involve casting the bitset's data to `[]u64` before encoding/decoding, but it has been a while since I worked on that code.
I added this note because we will need it for 32-bit compatibility (a requirement for the WASM work).

jorangreef replied:
Ah, I meant that, at present, an empty free set trailer still has a machine word in it, whereas, an empty manifest trailer is completely empty. So the encoding/decoding routines in the superblock are a little different IIRC.

sentientwaffle replied:
Ah, I didn't realize that!

jorangreef commented on src/vsr/superblock_client_table.zig:
```diff
@@ -45,7 +45,7 @@ pub const ClientTable = struct {
         assert(entries.capacity() >= config.clients_max);
 
-        const sorted = try allocator.alloc(*const Entry, entries.capacity());
+        const sorted = try allocator.alloc(*const Entry, config.clients_max);
```
I think that `capacity()` could in fact exceed `clients_max` due to hash table load factor? However, this shouldn't be necessary when it comes to sorting?

sentientwaffle replied:
`capacity()` can exceed `clients_max`, but `clients_max` is the maximum number of entries we will store in the hash map (and thus, the maximum that will need to be sorted). So this is (potentially) a slightly tighter bound.

jorangreef commented on src/vsr/superblock_manifest.zig:
```diff
@@ -25,8 +35,8 @@ pub const Manifest = struct {
-        block: u64,
-        entry: u32,
+        block: u64, // ManifestLog block address
+        entry: u32, // index within the ManifestLog Label/TableInfo arrays
```
```suggestion
        block: u64, // ManifestLog block address.
        entry: u32, // index within the ManifestLog Label/TableInfo arrays.
```

jorangreef reviewed (commented):
Awesome improvement!

sentientwaffle committed:
StateMachine: Fix commit_count_max configuration

- Move the state machine's storage configuration out of `config.zig` and into the `StateMachine`. The values of these configs are tightly coupled to the internal state machine logic. (For example, the `commit_count_max` can vary per-tree depending on how the field is updated).
- "Fix" the account's `commit_count_max`. The value was actually correct, but for a completely different reason than I thought. (This is now documented in the code).
- Fix the transfer's `commit_count_max` — transfers are immutable, so no multiplier is needed.
- Split `commit_count_max` configuration up by `Tree` (instead of just `Groove`, as it was before). Many fields are immutable (e.g. `ledger`), which gives them a tighter bound.

sentientwaffle committed:
LSM: Cleanup/clarification/fixes from code review

sentientwaffle committed:
SuperBlock: in-memory test

sentientwaffle committed:
SuperBlock: Fix client table encoding

sentientwaffle committed:
LSM: Use doNotOptimizeAway in benchmark

jorangreef commented on src/vsr/superblock_client_table.zig:
```diff
@@ -184,16 +184,16 @@ pub const ClientTable = struct {
         const headers = mem.bytesAsSlice(
             vsr.Header,
-            source[size .. entries_count * @sizeOf(vsr.Header)],
+            source[size..][0 .. entries_count * @sizeOf(vsr.Header)],
```
Good catch!

jorangreef commented on src/lsm/set_associative_cache.zig:
```diff
@@ -105,12 +106,18 @@ pub fn SetAssociativeCache(
         /// On cache write, entries are checked for occupancy (or eviction) beginning from the
         /// clock's position, wrapping around.
+        ///
+        /// The algorithm implemented is "CLOCK Nth-Chance" — each way has more than one bit,
+        /// to give ways more than one chance before eviction.
+        ///
+        /// * A similar algorithm called "RRIParoo" is described in
+        ///   "Kangaroo: Caching Billions of Tiny Objects on Flash".
+        /// * For more general information on CLOCK algorithms, see:
+        ///   https://en.wikipedia.org/wiki/Page_replacement_algorithm.
```
Great.

sentientwaffle committed:
LSM: "cache_count_max" → "cache_cardinality_max"

sentientwaffle committed:
VSR: Tick IO at top-level, Storage in replica

Reasoning:
- There is 1 storage : 1 replica ratio, but if you (for some reason) had multiple replicas in a process, they would all share an IO, and you wouldn't want to tick IO that more than once per replica tick.

jorangreef commented on src/main.zig:
```diff
@@ -149,7 +149,10 @@ const Command = struct {
         while (true) {
             replica.tick();
-            command.storage.tick();
+            command.io.tick() catch |err| {
+                log.warn("tick: {}", .{err});
+                std.debug.panic("io tick: {}", .{err});
+            };
             try command.io.run_for_ns(config.tick_ms * std.time.ns_per_ms);
```
This also lines up now with the next line, which is then to call `io.run_for_ns()`.

So we have the encapsulation right now!

jorangreef commented on src/storage.zig:
```diff
@@ -84,10 +84,8 @@ pub const Storage = struct {
-            std.debug.panic("storage tick: {}", .{err});
-        };
+        _ = storage;
+        // IO is ticked by the top-level (e.g. main.zig).
```
... since it may be shared by more than one replica.
// For example, when running in the simulator.

sentientwaffle replied:
I wrote `for example during tests.` instead of specifying the simulator, the avoid confusion because the simulator doesn't use IO.

jorangreef commented on src/config.zig:
```diff
@@ -49,22 +49,22 @@ pub const memory_size_max_default = 1024 * 1024 * 1024;
 /// The maximum number of two-phase transfers to store in memory:
 /// This impacts the amount of memory allocated at initialization by the server.
-pub const transfers_pending_max = transfers_max;
+pub const cache_transfers_pending_max = cache_transfers_max;
```
These values could be too high. For example, 100m transfers is 12.8 GiB.

They'll eventually be runtime known, instead of constants, but for now what do you think of going with 100k accounts (12.8 MiB) and 1m transfers (128 MiB)?

jorangreef commented on src/lsm/table_mutable.zig:
```diff
@@ -65,8 +65,8 @@ pub fn TableMutableType(comptime Table: type) type {
             // level getOrPut() API and manually overwrite the old key.
-            const gop = table.values.getOrPutAssumeCapacity(value.*);
-            gop.key_ptr.* = tombstone_from_key(key_from_value(value));
+            const upsert = table.values.getOrPutAssumeCapacity(value.*);
```
Lekker! ("good!") :)

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -397,33 +420,21 @@ pub fn GrooveType(
         /// sufficient to query this hashmap alone to know the state of the LSM trees.
         prefetch_objects: PrefetchObjects,
 
+        pub const IndexTreesOptions = IndexTreesOptions;
```
Is there a shadowing issue here with both names the same? Or am I seeing double?

sentientwaffle replied:
Nope, this is dead code :facepalm:. Good eye!

jorangreef commented on src/lsm/test.zig:
```diff
@@ -46,18 +29,13 @@ const Environment = struct {
-        },
-    };
+    const forest_options: Forest.Options = StateMachine.forest_options(.{
+        .lsm_forest_node_count = undefined, // ignored by StateMachine.forest_options()
```
`// Ignored by StateMachine.forest_options().`

jorangreef commented on src/state_machine.zig:
```diff
@@ -908,6 +887,97 @@ pub fn StateMachineType(comptime Storage: type) type {
+                                batch_accounts_max,
+                                2 * batch_transfers_max,
+                            ),
+                        },
```
This is so powerful. Nice work.

jorangreef replied:
Been having fun getting my head around the 2x multiplier for debits/credits the past few days. This is a great explanation.

sentientwaffle committed:
LSM: Cleanup; reduce cache sizes

Rename `cache_cardinality_max`, `prefetch_count_max` and `commit_count_max` to be explicit entries are what is being counted.
The default cache sizes were too high.
